File: scripts/convertpsl.py
#!/usr/bin/env python

# script based on
# https://github.com/adblockplus/buildtools/blob/d090e00610a58cebc78478ae33e896e6b949fc12/publicSuffixListUpdater.py

from __future__ import print_function

import json
import sys

def convert(psl_text):
    suffixes = {}

    for line in psl_text:
        line = line.rstrip()
        if line.startswith('//') or '.' not in line:
            continue
        if line.startswith('*.'):
            suffixes[line[2:]] = 2
        elif line.startswith('!'):
            suffixes[line[1:]] = 0
        else:
            suffixes[line] = 1

    return suffixes


if __name__ == '__main__':
    with open(sys.argv[1], 'r+') as f:
        psl = convert(f)
        f.seek(0)
        text = 'window.publicSuffixes = %s;' % (
            json.dumps(psl, sort_keys=True, indent=2, separators=(',', ': '))
        )
        print(text, file=f)
        f.truncate()

File: scripts/convertpsl.js
/* eslint-env node */

// script based on
// https://github.com/adblockplus/adblockpluscore/blob/d7efa99b47e6cefa9286d34ce80e2242d789a08b/build/updatepsl.js

const fs = require('fs');

function convert(data) {
  let psl = {};

  for (let line of data.split(/\r?\n/)) {
    if (line.startsWith("//") || !line.includes(".")) {
      continue;
    }

    let value = 1;
    line = line.replace(/\s+$/, "");

    if (line.startsWith("*.")) {
      line = line.slice(2);
      value = 2;
    } else if (line.startsWith("!")) {
      line = line.slice(1);
      value = 0;
    }

    psl[new URL("http://" + line).hostname] = value;
  }

  return psl;
}

if (require.main == module) {
  let filename = process.argv[2];

  let psl = convert(fs.readFileSync(filename, 'utf8'));

  let keys = Object.keys(psl).sort();
  fs.writeFileSync(
    filename,
    `window.publicSuffixes = ${JSON.stringify(psl, keys, 2)};\n`,
    'utf8');
}
